src/lib/runway_takeoff/RunwayTakeoff.cpp

```diff
@@ -0,0 +1,248 @@
+#include <mavlink/mavlink_log.h>
+#include <mathlib/mathlib.h>
+#include <math.h>
+
+namespace runwaytakeoff
+{
+
+RunwayTakeoff::RunwayTakeoff() :
+	SuperBlock(NULL, "RWTO"),
+	_state(),
+	_initialized(false),
+	_initialized_time(0),
+	_init_yaw(0),
+	_climbout(false),
+	_start_sp{},
+	_target_sp{},
+	_runway_takeoff_enabled(this, "TKOFF"),
+	_heading_mode(this, "HDG"),
+	_nav_alt(this, "NAV_ALT"),
+	_takeoff_throttle(this, "MAX_THR"),
+	_runway_pitch_sp(this, "PSP"),
+	_max_takeoff_pitch(this, "MAX_PITCH"),
+	_max_takeoff_roll(this, "MAX_ROLL"),
+	_min_airspeed_scaling(this, "AIRSPD_SCL"),
+	_airspeed_min(this, "FW_AIRSPD_MIN", false),
+	_climbout_diff(this, "FW_CLMBOUT_DIFF", false)
+{
+
+	updateParams();
+}
+
+RunwayTakeoff::~RunwayTakeoff()
+{
+
+}
+
+void RunwayTakeoff::init(float yaw)
+{
+	_init_yaw = yaw;
+	_initialized = true;
+	_state = RunwayTakeoffState::THROTTLE_RAMP;
+	_initialized_time = hrt_absolute_time();
+	_climbout = true;
+}
+
+void RunwayTakeoff::update(float airspeed, float alt_agl, int mavlink_fd)
+{
+
+	switch (_state) {
+	case RunwayTakeoffState::THROTTLE_RAMP:
+		if (hrt_elapsed_time(&_initialized_time) > 1e6) {
+			_state = RunwayTakeoffState::CLAMPED_TO_RUNWAY;
+		}
+
+		break;
+
+	case RunwayTakeoffState::CLAMPED_TO_RUNWAY:
+		if (airspeed > _airspeed_min.get() * _min_airspeed_scaling.get()) {
+			_state = RunwayTakeoffState::TAKEOFF;
+			mavlink_log_info(mavlink_fd, "#Takeoff airspeed reached");
+		}
+
+		break;
+
+	case RunwayTakeoffState::TAKEOFF:
+		if (alt_agl > _nav_alt.get()) {
+			_state = RunwayTakeoffState::CLIMBOUT;
+			mavlink_log_info(mavlink_fd, "#Climbout");
+		}
+
+		break;
+
+	case RunwayTakeoffState::CLIMBOUT:
+		if (alt_agl > _climbout_diff.get()) {
+			_climbout = false;
+			_state = RunwayTakeoffState::FLY;
+			mavlink_log_info(mavlink_fd, "#Navigating to waypoint");
+		}
+
+		break;
+
+	default:
+		return;
+	}
+}
+
+bool RunwayTakeoff::controlYaw()
+{
+	// keep controlling yaw directly until we start navigation
+	return _state < RunwayTakeoffState::CLIMBOUT;
+}
+
+float RunwayTakeoff::getPitch(float tecsPitch)
+{
+	if (_state <= RunwayTakeoffState::CLAMPED_TO_RUNWAY) {
+		return math::radians(_runway_pitch_sp.get());
+	}
+
+	return tecsPitch;
+}
+
+float RunwayTakeoff::getRoll(float navigatorRoll)
+{
+	// until we have enough ground clearance, set roll to 0
+	if (_state < RunwayTakeoffState::CLIMBOUT) {
+		return 0.0f;
+	}
+
+	// allow some roll during climbout
+	else if (_state < RunwayTakeoffState::FLY) {
+		return math::constrain(navigatorRoll,
+				       math::radians(-_max_takeoff_roll.get()),
+				       math::radians(_max_takeoff_roll.get()));
+	}
+
+	return navigatorRoll;
+}
+
+float RunwayTakeoff::getYaw(float navigatorYaw)
+{
+	return navigatorYaw;
+}
+
+float RunwayTakeoff::getThrottle(float tecsThrottle)
+{
+	switch (_state) {
+	case RunwayTakeoffState::THROTTLE_RAMP: {
+			float throttle = hrt_elapsed_time(&_initialized_time) / (float)2000000 *
```

Comment:
I think this magic number and the 1e6 in the switch case should be one const/define.